File: loading-button-android/src/main/java/br/com/simplepass/loading_button_lib/animatedDrawables/CircularAnimatedDrawable.java
package br.com.simplepass.loading_button_lib.animatedDrawables;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.AnimatorSet;
import android.animation.ValueAnimator;
import android.graphics.Canvas;
import android.graphics.ColorFilter;
import android.graphics.Paint;
import android.graphics.PixelFormat;
import android.graphics.Rect;
import android.graphics.RectF;
import android.graphics.drawable.Animatable;
import android.graphics.drawable.Drawable;
import android.view.View;
import android.view.animation.AccelerateDecelerateInterpolator;
import android.view.animation.Interpolator;
import android.view.animation.LinearInterpolator;

/**
 * Created by leandro on 5/31/16.
 */
public class CircularAnimatedDrawable extends Drawable implements Animatable {
    private ValueAnimator mValueAnimatorAngle;
    private ValueAnimator mValueAnimatorSweep;
    private AnimatorSet mAnimatorSet;
    private static final Interpolator ANGLE_INTERPOLATOR = new LinearInterpolator();
    private static final Interpolator SWEEP_INTERPOLATOR = new AccelerateDecelerateInterpolator();
    private static final int ANGLE_ANIMATOR_DURATION = 2000;
    private static final int SWEEP_ANIMATOR_DURATION = 700;
    private static final Float MIN_SWEEP_ANGLE = 50f;

    private final RectF fBounds = new RectF();
    private Paint mPaint;
    private View mAnimatedView;

    private float mBorderWidth;
    private float mCurrentGlobalAngle;
    private float mCurrentSweepAngle;
    private float mCurrentGlobalAngleOffset;

    private boolean mModeAppearing;
    private boolean mRunning;

    private boolean shouldDraw;

    /**
     *
     * @param view View to be animated
     * @param borderWidth The width of the spinning bar
     * @param arcColor The color of the spinning bar
     */
    public CircularAnimatedDrawable(View view, float borderWidth, int arcColor) {
        mAnimatedView = view;

        mBorderWidth = borderWidth;

        mPaint = new Paint();
        mPaint.setAntiAlias(true);
        mPaint.setStyle(Paint.Style.STROKE);
        mPaint.setStrokeWidth(borderWidth);
        mPaint.setColor(arcColor);

        setupAnimations();

        shouldDraw = true;

        mAnimatorSet = new AnimatorSet();
    }

    @Override
    protected void onBoundsChange(Rect bounds) {
        super.onBoundsChange(bounds);
        fBounds.left = bounds.left + mBorderWidth / 2f + .5f;
        fBounds.right = bounds.right - mBorderWidth / 2f - .5f;
        fBounds.top = bounds.top + mBorderWidth / 2f + .5f;
        fBounds.bottom = bounds.bottom - mBorderWidth / 2f - .5f;
    }

    public void setLoadingBarColor(int color) {
        mPaint.setColor(color);
    }

    /**
     * Start the animation
     */
    @Override
    public void start() {
        if (isRunning()) {
            return;
        }

        mRunning = true;

        mAnimatorSet.playTogether(mValueAnimatorAngle, mValueAnimatorSweep);
        mAnimatorSet.start();
    }

    /**
     * Stops the animation
     */
    @Override
    public void stop() {
        if (!isRunning()) {
            return;
        }

        mRunning = false;
        mAnimatorSet.cancel();
    }

    /**
     * Method the inform if the animation is in process
     *
     * @return
     */
    @Override
    public boolean isRunning() {
        return mRunning;
    }

    /**
     * Method called when the drawable is going to draw itself.
     * @param canvas
     */
    @Override
    public void draw(Canvas canvas) {
        float startAngle = mCurrentGlobalAngle - mCurrentGlobalAngleOffset;
        float sweepAngle = mCurrentSweepAngle;

        if (!mModeAppearing) {
            startAngle = startAngle + sweepAngle;
            sweepAngle = 360 - sweepAngle - MIN_SWEEP_ANGLE;
        } else {
            sweepAngle += MIN_SWEEP_ANGLE;
        }

        canvas.drawArc(fBounds, startAngle, sweepAngle, false, mPaint);
    }

    @Override
    public void setAlpha(int alpha) {
        mPaint.setAlpha(alpha);
    }

    @Override
    public void setColorFilter(ColorFilter colorFilter) {
        mPaint.setColorFilter(colorFilter);
    }

    @Override
    public int getOpacity() {
        return PixelFormat.TRANSPARENT;
    }

    /**
     * Set up all the animations. There are two animation: Global angle animation and sweep animation.
     */
    private void setupAnimations() {
        mValueAnimatorAngle = ValueAnimator.ofFloat(0, 360f);
        mValueAnimatorAngle.setInterpolator(ANGLE_INTERPOLATOR);
        mValueAnimatorAngle.setDuration(ANGLE_ANIMATOR_DURATION);
        mValueAnimatorAngle.setRepeatCount(ValueAnimator.INFINITE);
        mValueAnimatorAngle.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                mCurrentGlobalAngle = (float)animation.getAnimatedValue();
            }
        });

        mValueAnimatorSweep = ValueAnimator.ofFloat(0, 360f - 2 * MIN_SWEEP_ANGLE);
        mValueAnimatorSweep.setInterpolator(SWEEP_INTERPOLATOR);
        mValueAnimatorSweep.setDuration(SWEEP_ANIMATOR_DURATION);
        mValueAnimatorSweep.setRepeatCount(ValueAnimator.INFINITE);
        mValueAnimatorSweep.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationRepeat(Animator animation) {
                toggleAppearingMode();
                shouldDraw = false;
            }
        });

        mValueAnimatorSweep.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                mCurrentSweepAngle = (float) animation.getAnimatedValue();

                if (mCurrentSweepAngle < 5) {
                    shouldDraw = true;
                }

                if (shouldDraw) {
                    mAnimatedView.invalidate();
                }
            }
        });

    }

    /**
     * This method is called in every repetition of the animation, so the animation make the sweep
     * growing and then make it shirinking.
     */
    private void toggleAppearingMode() {
        mModeAppearing = !mModeAppearing;

        if (mModeAppearing) {
            mCurrentGlobalAngleOffset = (mCurrentGlobalAngleOffset + MIN_SWEEP_ANGLE * 2) % 360;
        }
    }


    public void dispose() {
        if (mValueAnimatorAngle != null) {
            mValueAnimatorAngle.end();
            mValueAnimatorAngle.removeAllUpdateListeners();
            mValueAnimatorAngle.cancel();
        }

        mValueAnimatorAngle = null;

        if (mValueAnimatorSweep != null) {
            mValueAnimatorSweep.end();
            mValueAnimatorSweep.removeAllUpdateListeners();
            mValueAnimatorSweep.cancel();
        }

        mValueAnimatorSweep = null;

        if (mAnimatorSet != null) {
            mAnimatorSet.end();
            mAnimatorSet.cancel();
        }
    }
}

File: loading-button-android/src/main/java/br/com/simplepass/loading_button_lib/animatedDrawables/CircularAnimatedDrawable.kt
package br.com.simplepass.loading_button_lib.animatedDrawables

import android.animation.*
import android.graphics.*
import android.graphics.drawable.Animatable
import android.graphics.drawable.Drawable
import android.view.View
import android.view.animation.AccelerateDecelerateInterpolator
import android.view.animation.LinearInterpolator

const val MIN_PROGRESS = 0F
const val MAX_PROGRESS = 100F
private const val ANGLE_ANIMATOR_DURATION = 2000L
private const val SWEEP_ANIMATOR_DURATION = 700L
private const val PROGRESS_ANIMATOR_DURATION = 200L
private const val MIN_SWEEP_ANGLE = 50f

class CircularAnimatedDrawable(
    private val animatedView: View,
    private val borderWidth: Float,
    arcColor: Int
) : Drawable(), Animatable {

    private val fBounds = RectF()
    private val paint = Paint().apply {
        isAntiAlias = true
        style = Paint.Style.STROKE
        strokeWidth = borderWidth
        color = arcColor
    }

    private var currentGlobalAngle: Float = 0F
    private var currentSweepAngle: Float = 0F
    private var currentGlobalAngleOffset: Float = 0F

    private var modeAppearing: Boolean = false

    private var shouldDraw: Boolean = true
    private var shownProgress: Float = 0F

    private var progressType = ProgressType.INDETERMINATE
    private val determinateInterpolator = AccelerateDecelerateInterpolator()

    private val indeterminateAnimator = AnimatorSet().apply {
        playTogether(
            angleValueAnimator(LinearInterpolator()),
            sweepValueAnimator(determinateInterpolator)
        )
    }

    private fun toggleSweep() {
        modeAppearing = !modeAppearing

        if (modeAppearing) {
            currentGlobalAngleOffset = (currentGlobalAngleOffset + MIN_SWEEP_ANGLE * 2) % 360
        }
    }

    private fun angleValueAnimator(interpolator: TimeInterpolator): ValueAnimator =
        ValueAnimator.ofFloat(0F, 360F).apply {
            setInterpolator(interpolator)
            duration = ANGLE_ANIMATOR_DURATION
            repeatCount = ValueAnimator.INFINITE

            addUpdateListener { animation -> currentGlobalAngle = animation.animatedValue as Float }
        }

    private fun sweepValueAnimator(interpolator: TimeInterpolator): ValueAnimator =
        ValueAnimator.ofFloat(0F, 360F - 2 * MIN_SWEEP_ANGLE).apply {
            setInterpolator(interpolator)
            duration = SWEEP_ANIMATOR_DURATION
            repeatCount = ValueAnimator.INFINITE

            addUpdateListener { animation ->
                currentSweepAngle = animation.animatedValue as Float

                if (currentSweepAngle < 5) {
                    shouldDraw = true
                }

                if (shouldDraw) {
                    animatedView.invalidate()
                }
            }

            addListener(object : AnimatorListenerAdapter() {
                override fun onAnimationRepeat(animation: Animator) {
                    toggleSweep()
                    shouldDraw = false
                }
            })
        }

    private fun disposeAnimator(animator: Animator) {
        animator.run {
            end()
            removeAllListeners()
            cancel()
        }
    }

    private fun getAngles(): Pair<Float, Float> =
        when (progressType) {
            ProgressType.DETERMINATE   -> {
                -90f to shownProgress
            }
            ProgressType.INDETERMINATE -> {
                if (modeAppearing) {
                    (currentGlobalAngle - currentGlobalAngleOffset) to currentSweepAngle + MIN_SWEEP_ANGLE
                } else {
                    (currentGlobalAngle - currentGlobalAngleOffset + currentSweepAngle) to
                        360f - currentSweepAngle - MIN_SWEEP_ANGLE
                }
            }
        }

    fun setLoadingBarColor(color: Int) {
        paint.color = color
    }

    override fun onBoundsChange(bounds: Rect) {
        super.onBoundsChange(bounds)

        fBounds.run {
            left = bounds.left.toFloat() + borderWidth / 2F + .5F
            right = bounds.right.toFloat() - borderWidth / 2F - .5F
            top = bounds.top.toFloat() + borderWidth / 2F + .5F
            bottom = bounds.bottom.toFloat() - borderWidth / 2F - .5F
        }
    }

    override fun isRunning(): Boolean = indeterminateAnimator.isRunning

    override fun start() {
        if (isRunning) {
            return
        }

        indeterminateAnimator.start()
    }

    override fun stop() {
        if (!isRunning) {
            return
        }

        indeterminateAnimator.end()
    }

    override fun draw(canvas: Canvas) {
        val (startAngle, sweepAngle) = getAngles()
        canvas.drawArc(fBounds, startAngle, sweepAngle, false, paint)
    }

    override fun setAlpha(alpha: Int) {
        paint.alpha = alpha
    }

    override fun getOpacity(): Int = PixelFormat.TRANSPARENT

    override fun setColorFilter(colorFilter: ColorFilter?) {
        paint.colorFilter = colorFilter
    }

    fun dispose() {
        disposeAnimator(indeterminateAnimator)
    }

    fun setProgress(newProgress: Float) {
        if (progressType == ProgressType.INDETERMINATE) {
            stop()
            progressType = ProgressType.DETERMINATE
        }

        if (shownProgress == newProgress) {
            return
        }

        shownProgress = when {
            newProgress > MAX_PROGRESS -> MAX_PROGRESS
            newProgress < MIN_PROGRESS -> MIN_PROGRESS
            else                       -> newProgress
        }

        invalidateSelf()
    }
}
